### ISSUE TITLE
Highlight issue with hack files without opening markup tag.

### ISSUE THREAD

**Issue by azjezz** [github/linguist#4466]
Hack files that don't include `<?hh` opening tag are not being highlighted.

## Preliminary Steps

Please confirm you have...
- [x] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
- [x] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
- [x] considered implementing an [override](https://github.com/github/linguist#overrides),
- [x] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).

## Problem Description

Hack files that don't include `<?hh` opening tag are not being highlighted using the `hack` syntax. instead it uses HTML highlighting.
if `<?` is encountered, the file will be highlighted starting from the `<?` position, even though that `<?` is not an opening tag.

Example of hack file not being highlighted : 
- https://github.com/nuxed/framework/blob/master/src/Nuxed/Http/Server/Middleware/HostMiddlewareDecorator.hack

Example of HTML highlighting :
- https://github.com/nuxed/framework/blob/master/src/Nuxed/Asset/Packages.hack#L9
- https://github.com/nuxed/framework/blob/async-io/src/Nuxed/Http/Session/Session.hack#L145

Example of a non-opening `<?` tag : 
- https://github.com/azjezz/mutex/blob/master/src/Memory.hack#L8

Minimal example : 
```
use namespace HH\Lib\SecureRandom;
                             // v-- highlighting starts from here
async function foo(): Awaitable<?string> {
  $bar = SecureRandom\int(0, 10);
  if ($bar > 8) {
    return 'baz';
  }

  return null;
}
```

### URL of the affected repository:

- https://github.com/nuxed/framework
- https://github.com/azjezz/mutex
- https://github.com/hhvm/hhvm-autoload
- https://github.com/hhvm/definition-finder

and many other projects that use hack with `.hack` file extension.

### Last modified on:
2019-03-19

**Comment by pchaigno** [comment]
This looks like a problem with the grammar GitHub uses to highlight Hack files, not with Linguist. Linguist does tell GitHub to use https://github.com/textmate/php.tmbundle as the grammar, so maybe you know a better grammar we could tell GitHub to use (Sublime Text, Atom, or TextMate package)?

**Comment by azjezz** [comment]
the @slackHQ [hack extension](https://github.com/slackhq/vscode-hack) for vscode does well highlighting hack files, but i don't know if it could be used here.

syntax file : https://github.com/slackhq/vscode-hack/blob/master/syntaxes/hack.json

( the issue of linguist not detecting `.hack` as hack, already has a PR : #4354 )

**Comment by pchaigno** [comment]
It looks like the Atom format and should work. Do you want to open a pull request? There are guidelines at [CONTRIBUTING.md](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#changing-the-source-of-a-syntax-highlighting-grammar).

**Comment by azjezz** [comment]
i had an issue adding the grammar : 

```
  > Status: Downloaded newer image for linguist/grammar-compiler:latest
  > The new grammar repository `vendor/grammars/vscode-hack` (from https://github.com/slackhq/vscode-hack) contains 1 errors:
  >     - Grammar conversion failed. File `syntaxes/hack.json` failed to parse: 1 error(s) decoding:
  > 
  > * 'Repository[type-annotation].Patterns[3].Patterns[0].EndCaptures[1]' expected a map, got 'string'
  > 
  > failed to compile the given grammar
Command failed. Aborting.
```

maybe you could take a look at it ?

**Comment by pchaigno** [comment]
There's a syntax error in the source file of the grammar. [That's how it should be](https://github.com/pchaigno/vscode-hack/commit/e7124629c89190d4b75cefe4538281ffeca18205). However, that file is generated from [facebookarchive/atom-ide-hack](https://github.com/facebookarchive/atom-ide-hack/blob/master/grammars/hack.cson) which is itself archived and read-only. I'm not sure where I should send a patch... Maybe there's a better grammar somewhere?

**Comment by azjezz** [comment]
send the path to slack repository, the atom extension is archived and vscode is now recommended, few changes have been made since the original `syntax.json` was generated.

**Comment by PranayAgarwal** [comment]
That's correct. `hack.json` in the vscode-hack is now the definitive version. I'll update the project readme to reflect that and remove references to the atom-ide project.

**Comment by azjezz** [comment]
fixed in #4544
